src/org/labkey/remoteapi/assay/Protocol.java

```diff
@@ -0,0 +1,415 @@
+            _editableRuns = (Boolean)json.get("editableRuns");
+        if (json.containsKey("saveScriptFiles"))
+            _saveScriptFiles = (Boolean)json.get("saveScriptFiles");
+        if (json.containsKey("qcEnabled"))
+            _qcEnabled = (Boolean)json.get("qcEnabled");
+        if (json.containsKey("allowQCStates"))
+            _allowQCStates = (Boolean)json.get("allowQCStates");
+        if (json.containsKey("allowSpacesInPath"))
+            _allowSpacesInPath = (Boolean)json.get("allowSpacesInPath");
+        if (json.containsKey("allowTransformationScript"))
+            _allowTransformationScript = (Boolean)json.get("allowTransformationScript");
+        if (json.containsKey("autoCopyTargetContainerId"))
+            _autoCopyTargetContainerId = (String)json.get("autoCopyTargetContainerId");
+
+        if (json.get("availableDetectionMethods") instanceof JSONArray)
+        {
+            for (Object detectionMethod : (JSONArray)json.get("availableDetectionMethods"))
+                _availableDetectionMethods.add((String)detectionMethod);
+        }
+        if (json.containsKey("selectedDetectionMethod"))
+            _selectedDetectionMethod = (String)json.get("selectedDetectionMethod");
+
+        if (json.containsKey("availableMetadataInputFormats"))
+        {
+            _availableMetadataInputFormats = (HashMap<String,String>)json.get("availableMetadataInputFormats");
+        }
+        if (json.containsKey("selectedMetadataInputFormat"))
+            _selectedMetadataInputFormat = (String)json.get("selectedMetadataInputFormat");
+
+        if (json.get("availablePlateTemplates") instanceof JSONArray)
+        {
+            for (Object plateTemplate : (JSONArray)json.get("availablePlateTemplates"))
+                _availablePlateTemplates.add((String)plateTemplate);
+        }
+        if (json.containsKey("selectedPlateTemplate"))
+            _selectedPlateTemplate = (String)json.get("selectedPlateTemplate");
+
+        if (json.get("protocolTransformScripts") instanceof JSONArray)
+        {
+            for (Object transformScript : (JSONArray)json.get("protocolTransformScripts"))
+                _protocolTransformScripts.add((String)transformScript);
+        }
+        if (json.containsKey("protocolParameters"))
+        {
+            _protocolParameters = (HashMap<String,String>)json.get("protocolParameters");
+        }
+    }
+
+    public JSONObject toJSONObject()
+    {
+        JSONObject result = new JSONObject();
+        result.put("protocolId", _protocolId);
+        result.put("name", _name);
+        result.put("description", _description);
+        result.put("providerName", _providerName);
+        JSONArray domains = new JSONArray();
+        result.put("domains", domains);
+        for (Domain domain : _domains)
+            domains.add(domain.toJSONObject(true));
+
+        if (_allowBackgroundUpload != null)
+            result.put("allowBackgroundUpload", _allowBackgroundUpload);
```

Comment:
I believe that this toJSONObject method is only called in the save case. If so, I think technically we could leave off the various "allow" related parameters since those are just informational coming back from the GetProtocolCommand and not something that can be set / saved. @labkey-klum thoughts?

Reply:
Agreed. Chris, we talked about these "allow" properties last week. They're read-only.

Reply:
Okay, I'll take those out but add a note on which ones are absent, and why